src/codegen/helpers.rs

```diff
@@ -1,66 +1,62 @@
 //! Helpers for code generation that don't need macro expansion.
 
-use aster;
 use ir::layout::Layout;
-use syntax::ast;
-use syntax::ptr::P;
-
+use quote;
 
 pub mod attributes {
-    use aster;
-    use syntax::ast;
-
-    pub fn allow(which_ones: &[&str]) -> ast::Attribute {
-        aster::AstBuilder::new()
-            .attr()
-            .list("allow")
-            .words(which_ones)
-            .build()
-    }
+    use quote;
 
-    pub fn repr(which: &str) -> ast::Attribute {
-        aster::AstBuilder::new()
-            .attr()
-            .list("repr")
-            .words(&[which])
-            .build()
+    pub fn repr(which: &str) -> quote::Tokens {
+        let which = quote::Ident::new(which);
+        quote! {
+            #[repr( #which )]
+        }
     }
 
-    pub fn repr_list(which_ones: &[&str]) -> ast::Attribute {
-        aster::AstBuilder::new()
-            .attr()
-            .list("repr")
-            .words(which_ones)
-            .build()
+    pub fn repr_list(which_ones: &[&str]) -> quote::Tokens {
+        let which_ones = which_ones.iter().cloned().map(quote::Ident::new);
+        quote! {
+            #[repr( #( #which_ones ),* )]
+        }
     }
 
-    pub fn derives(which_ones: &[&str]) -> ast::Attribute {
-        aster::AstBuilder::new()
-            .attr()
-            .list("derive")
-            .words(which_ones)
-            .build()
+    pub fn derives(which_ones: &[&str]) -> quote::Tokens {
+        let which_ones = which_ones.iter().cloned().map(quote::Ident::new);
+        quote! {
+            #[derive( #( #which_ones ),* )]
+        }
     }
 
-    pub fn inline() -> ast::Attribute {
-        aster::AstBuilder::new().attr().word("inline")
+    pub fn inline() -> quote::Tokens {
+        quote! {
+            #[inline]
+        }
     }
 
-    pub fn doc(comment: String) -> ast::Attribute {
-        aster::AstBuilder::new().attr().doc(&*comment)
+    pub fn doc(comment: String) -> quote::Tokens {
+        // Doc comments are already preprocessed into nice `///` formats by the
+        // time they get here. Just make sure that we have newlines around it so
+        // that nothing else gets wrapped into the comment.
+        let mut tokens = quote! {};
+        tokens.append("\n");
+        tokens.append(comment);
+        tokens.append("\n");
+        tokens
     }
 
-    pub fn link_name(name: &str) -> ast::Attribute {
-        aster::AstBuilder::new()
-            .attr()
-            .name_value("link_name")
-            .str(name)
+    pub fn link_name(name: &str) -> quote::Tokens {
+        // This is ridiculous, but rustfmt doesn't seem to be formatting
+        // `link_name` attributes very well, so we jump through these formatting
+        // hoops to make it work.
```

Comment:
Can you file an issue upstream and reference it here?

Reply:
I've filed a handful of bugs upstream, and they've all been fixed now, so this comment might actually be out of date. I'll double check.